File: problems-originals/0601-0700/0692_top-k-frequent-words/solution_full_sort.ts
function topKFrequent(words: string[], k: number): string[] {
    // One counting pass over the array.
    const counts = new Map<string, number>();
    for (const w of words) {
        counts.set(w, (counts.get(w) || 0) + 1);
    }
    const ranked = Array.from(counts);
    // Sort every unique word under the statement's total order — count
    // descending, then word ascending — and keep the first k. The
    // comparator is explicit about both keys — never the default
    // lexicographic sort — and fully orders every pair, so no
    // sort-stability assumption can leak in.
    ranked.sort((a, b) => (a[1] !== b[1] ? b[1] - a[1] : a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0));
    return ranked.slice(0, k).map(([word]) => word);
}

File: problems-adapt/0601-0700/0692_ranked-word-frequencies/solution_full_sort.rs
use std::collections::HashMap;

impl Solution {
    pub fn rank_word_frequencies(words: Vec<String>, k: i32) -> Vec<String> {
        // One counting pass over the array.
        let mut counts: HashMap<&str, i32> = HashMap::new();
        for w in &words {
            *counts.entry(w.as_str()).or_insert(0) += 1;
        }
        let mut ranked: Vec<(i32, &str)> = counts.into_iter().map(|(word, count)| (count, word)).collect();
        // Sort every unique word under the statement's total order — count
        // descending, then word ascending — and keep the first k.
        ranked.sort_by(|a, b| b.0.cmp(&a.0).then_with(|| a.1.cmp(&b.1)));
        ranked
            .into_iter()
            .take(k as usize)
            .map(|(_, word)| word.to_string())
            .collect()
    }
}
